mmv1/third_party/terraform/utils/framework_location.go

```diff
@@ -0,0 +1,99 @@
+package google
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+type LocationDescriber interface {
+	getLocationDescription(providerConfig *frameworkProvider) LocationDescription
+}
+
+type LocationDescription struct {
+	LocationSchemaField types.String
+	RegionSchemaField   types.String
+	ZoneSchemaField     types.String
+
+	ResourceLocation types.String
+	ResourceRegion   types.String
+	ResourceZone     types.String
+
+	ProviderRegion types.String
+	ProviderZone   types.String
+}
+
+func (ld *LocationDescription) getLocation() (types.String, error) {
+	// Location from resource config
+	if !ld.ResourceLocation.IsNull() && !ld.ResourceLocation.IsUnknown() {
+		return ld.ResourceLocation, nil
+	}
+
+	// Location from region in resource config
+	if !ld.ResourceRegion.IsNull() && !ld.ResourceRegion.IsUnknown() {
+		return ld.ResourceRegion, nil
+	}
+
+	// Location from zone in resource config
+	if !ld.ResourceZone.IsNull() && !ld.ResourceZone.IsUnknown() {
+		location := GetResourceNameFromSelfLink(ld.ResourceZone.ValueString()) // Zone could be a self link
+		return types.StringValue(location), nil
+	}
+
+	// Location from zone in provider config
+	if !ld.ProviderZone.IsNull() {
+		return ld.ProviderZone, nil
+	}
+
+	return types.StringNull(), nil
+}
+
+func (ld *LocationDescription) getRegion() (types.String, error) {
+	// Region from resource config
+	if !ld.ResourceRegion.IsNull() && !ld.ResourceRegion.IsUnknown() {
+		region := GetResourceNameFromSelfLink(ld.ResourceRegion.ValueString()) // Region could be a self link
+		return types.StringValue(region), nil
+	}
+	// Region from zone in resource config
+	if !ld.ResourceZone.IsNull() && !ld.ResourceZone.IsUnknown() {
+		region := getRegionFromZone(ld.ResourceZone.ValueString())
```

Comment:
do we need to worry about zone being a self link here?

Reply:
(FWIW it looks like this matches the current behavior: https://github.com/hashicorp/terraform-provider-google/blob/15b2492b7192e4dcf03db4796080bf8a00239d3f/google/field_helpers.go#L386 but I recognize that doesn't answer the question)

Reply:
This is addressed in a test case and there's currently parity with how the previous, SDK function handles a self link. 
I'd be interested in identifying ways to make the logic more consistent, as I agree that it's odd that zones as self links are accommodated in some places but not others, but also I don't think it's a good idea to make the function _more_ permissive than it already is

Reply:
+1 on standardizing this, if we can find a way to do it without breaking users (or maybe it could be a target for the next major release?)

Reply:
Issue here: https://github.com/hashicorp/terraform-provider-google/issues/14503